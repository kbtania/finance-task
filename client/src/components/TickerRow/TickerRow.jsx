import React, {useEffect, useState} from 'react';

import styles from './TickerRow.module.css';
import {useDispatch, useSelector} from "react-redux";
import {tickersSelector} from "../../store/tickers/selector";
import {getChosenTicker, getTickersError} from "../../store/tickers/actionCreators";

const TickerRow = ({tickerItem}) => {
    const { ticker, price} = tickerItem;
    const [trendPrice, setTrendPrice] = useState(null);
    const { previousTickers, chosenTicker } = useSelector(tickersSelector);


    const dispatch = useDispatch();

    useEffect(() => {
        previousTickers.map((previousData) => {
            if (previousData.ticker === ticker &&
                Number(previousData.price) > Number(price)) {
                setTrendPrice(false);
            } if (previousData.ticker === ticker &&
                Number(previousData.price) < Number(price)) {
                setTrendPrice(true);
            }
            return previousData;
        });
    }, [previousTickers, price, ticker]);

    function handleChoice() {
        dispatch(getTickersError)
        dispatch(getChosenTicker(ticker));
    }
    return (
        <tr onClick={handleChoice} className="tickerRow" className={chosenTicker === ticker ? styles.selected: styles.row}>
            <td><div className={styles.ticker}>{tickerItem.ticker}</div></td>
            <td><div>{tickerItem.exchange}</div></td>
            <td>
                <div className={trendPrice ? styles.positive : styles.negative}>
                    {trendPrice ? <span className={styles.arrowUp}>&#9650;</span> : <span className={styles.arrowDown}>&#9660;</span>}
                    {tickerItem.price}
                </div></td>
            <td><div>{tickerItem.change}</div></td>
            <td>
                <div>
                    {tickerItem.change_percent}
                </div>
            </td>
            <td><div>{tickerItem.dividend}</div></td>
            <td><div>{tickerItem.yield}</div></td>
        </tr>
    );
};

export default TickerRow;