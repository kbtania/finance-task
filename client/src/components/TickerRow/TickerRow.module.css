.row {
    cursor: pointer;
}
.row:hover {
    background-color: #EEEEEE;
    transition: 0.5s ease;
}
.positive {
    background-color: #B1D7B4;
    padding: 5px;
    border-radius: 10px;
    text-align: center;
}

.negative {
    background-color: #FAD4D4;
    padding: 5px;
    border-radius: 10px;
    text-align: center;
}

.arrowUp {
    color: green;
    margin-right: 5px;
}
.arrowDown {
    color: red;
    margin-right: 5px;
}

.ticker {
    background-color: #2C3333;
    color: #E7F6F2;
    text-align: center;
    padding: 5px;
    border-radius: 10px;
}

.selected {
    cursor: pointer;
    background-color: #EEEEEE;
}