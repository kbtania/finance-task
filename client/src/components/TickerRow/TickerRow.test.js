import {render} from "@testing-library/react";
// import {useSelector} from "react-redux";
import * as reduxHooks from 'react-redux';

import TickerRow from "./TickerRow";

jest.mock('react-redux');

let item = {
    ticker: 'AAPL',
    exchange: 'NASDAQ',
    price: 105.89,
    change: 178.23,
    change_percent: 0.21,
    dividend: 1.64,
    yield: 1.56
}
let mockStore = {
    currentTickers: [
        {
            ticker: 'AAPL',
            exchange: 'NASDAQ',
            price: 105.89,
            change: 178.23,
            change_percent: 0.21,
            dividend: 1.64,
            yield: 1.56
        }
    ],
    previousTickers: [
        {
            ticker: 'AAPL',
            exchange: 'NASDAQ',
            price: 105.89,
            change: 178.23,
            change_percent: 0.21,
            dividend: 1.64,
            yield: 1.56
        }
    ],
    error: false,
    chosenTicker: 'AAPL'
}
describe('TickerRow component', () => {
    it('should create list with tickers', () => {
        jest.spyOn(reduxHooks, 'useSelector').mockReturnValue(mockStore)
        const component = render(<TickerRow tickerItem={item}/>);
        expect(component).toMatchSnapshot();
    })
})