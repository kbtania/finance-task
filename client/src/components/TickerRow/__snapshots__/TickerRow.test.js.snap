// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TickerRow component should create list with tickers 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <tr
        class="selected"
      >
        <td>
          <div
            class="ticker"
          >
            AAPL
          </div>
        </td>
        <td>
          <div>
            NASDAQ
          </div>
        </td>
        <td>
          <div
            class="negative"
          >
            <span
              class="arrowDown"
            >
              ▼
            </span>
            105.89
          </div>
        </td>
        <td>
          <div>
            178.23
          </div>
        </td>
        <td>
          <div>
            0.21
          </div>
        </td>
        <td>
          <div>
            1.64
          </div>
        </td>
        <td>
          <div>
            1.56
          </div>
        </td>
      </tr>
    </div>
  </body>,
  "container": <div>
    <tr
      class="selected"
    >
      <td>
        <div
          class="ticker"
        >
          AAPL
        </div>
      </td>
      <td>
        <div>
          NASDAQ
        </div>
      </td>
      <td>
        <div
          class="negative"
        >
          <span
            class="arrowDown"
          >
            ▼
          </span>
          105.89
        </div>
      </td>
      <td>
        <div>
          178.23
        </div>
      </td>
      <td>
        <div>
          0.21
        </div>
      </td>
      <td>
        <div>
          1.64
        </div>
      </td>
      <td>
        <div>
          1.56
        </div>
      </td>
    </tr>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
